source/extensions/filters/network/redis_proxy/conn_pool_impl.cc

```diff
@@ -206,24 +206,57 @@ PoolRequest* InstanceImpl::makeRequest(const std::string& hash_key, const RespVa
 
 InstanceImpl::ThreadLocalPool::ThreadLocalPool(InstanceImpl& parent, Event::Dispatcher& dispatcher,
                                                const std::string& cluster_name)
-    : parent_(parent), dispatcher_(dispatcher), cluster_(parent_.cm_.get(cluster_name)) {
+    : parent_(parent), dispatcher_(dispatcher), cluster_name_(cluster_name) {
 
-  // TODO(mattklein123): Redis is not currently safe for use with CDS. In order to make this work
-  //                     we will need to add thread local cluster removal callbacks so that we can
-  //                     safely clean things up and fail requests.
-  ASSERT(!cluster_->info()->addedViaApi());
-  local_host_set_member_update_cb_handle_ = cluster_->prioritySet().addMemberUpdateCb(
+  cluster_update_handle_ = parent_.cm_.addThreadLocalClusterUpdateCallbacks(*this);
+  Upstream::ThreadLocalCluster* cluster = parent_.cm_.get(cluster_name_);
+  if (cluster != nullptr) {
+    onClusterAddOrUpdate(*cluster);
+  }
+}
+
+InstanceImpl::ThreadLocalPool::~ThreadLocalPool() {
+  if (host_set_member_update_cb_handle_ != nullptr) {
+    host_set_member_update_cb_handle_->remove();
+  }
+  while (!client_map_.empty()) {
+    client_map_.begin()->second->redis_client_->close();
+  }
+}
+
+void InstanceImpl::ThreadLocalPool::onClusterAddOrUpdate(Upstream::ThreadLocalCluster& cluster) {
+  if (cluster.info()->name() != cluster_name_) {
+    return;
+  }
+
+  if (cluster_ != nullptr) {
+    // Treat an update as a removal followed by an add.
+    onClusterRemoval(cluster_name_);
+  }
+
+  ASSERT(cluster_ == nullptr);
+  cluster_ = &cluster;
+  ASSERT(host_set_member_update_cb_handle_ == nullptr);
+  host_set_member_update_cb_handle_ = cluster_->prioritySet().addMemberUpdateCb(
       [this](uint32_t, const std::vector<Upstream::HostSharedPtr>&,
              const std::vector<Upstream::HostSharedPtr>& hosts_removed) -> void {
         onHostsRemoved(hosts_removed);
       });
 }
 
-InstanceImpl::ThreadLocalPool::~ThreadLocalPool() {
-  local_host_set_member_update_cb_handle_->remove();
+void InstanceImpl::ThreadLocalPool::onClusterRemoval(const std::string& cluster_name) {
+  if (cluster_name != cluster_name_) {
+    return;
+  }
+
+  // Treat cluster removal as a removal of all hosts. Close all connections and fail all pending
+  // requests.
```

Comment:
When is this callback invoked? When I change cluster settings or when I add/remove hosts to the cluster via EDS?
Secondly, what happens if I have a cluster with logical or strict dns/static endpoints, and I update the cluster with more dns/static endpoints? Will it cause all ongoing connections to be dropped without going through the listener drain thingy?

Long story short, if redis connections are typically short lived transient connections, then this model of update (delete+add) may not have much effect. But if they are long lived sessions, this is going to be pretty annoying.

I do realize that the current redis filter does more than just load balancing or dumb passthrough. So, this might make sense. But I wonder if it makes sense to have a redis filter in simple passthrough mode like Mongo, where the filter extracts stats, and forwards the commands to one of the upstream hosts. [Atleast in Istio, this is what people are most likely to use to start with].

Reply:
> When is this callback invoked? When I change cluster settings or when I add/remove hosts to the cluster via EDS?

Only when cluster settings are changed (rare). EDS updates are seamless like all other host update changes.

> Secondly, what happens if I have a cluster with logical or strict dns/static endpoints, and I update the cluster with more dns/static endpoints? Will it cause all ongoing connections to be dropped without going through the listener drain thingy?

Yes, this counts as a cluster config change, everything will be dropped.

> Long story short, if redis connections are typically short lived transient connections, then this model of update (delete+add) may not have much effect. But if they are long lived sessions, this is going to be pretty annoying.

It's only annoying if you keep changing cluster settings, which doesn't make any sense in any normal usage scenario IMO.

> But I wonder if it makes sense to have a redis filter in simple passthrough mode like Mongo, where the filter extracts stats, and forwards the commands to one of the upstream hosts. [Atleast in Istio, this is what people are most likely to use to start with].

This is definitely possible but we should track that feature request separately as it's really a totally separate use case.